feat: update `keychain::Balance` to use `bitcoin::Amount`

oleonardolima opened the pull request:
fixes #823 

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Description

It's being used on `Balance`, and throughout the code, an `u64` represents the amount, which relies on the user to infer its sats, not millisats, or any other representation.

It updates the usage of `u64` on `Balance`, and other APIs:
- `TxParams::add_recipient`
- `KeyChainTxOutIndex::sent_and_received`, `KeyChainTxOutIndex::net_value`
-  `SpkTxOutIndex::sent_and_received`, `SpkTxOutIndex::net_value`

<!-- Describe the purpose of this PR, what's being adding and/or fixed -->

### Notes to the reviewers

<!-- In this section you can include notes directed to the reviewers, like explaining why some parts
of the PR were done in a specific way -->

It updates some of the APIs to expect the `bitcoin::Amount`, but it does not update internal usage of u64, such as `TxParams` still expects and uses `u64`, please see the PR comments for related discussion.

### Changelog notice

<!-- Notice the release manager should include in the release tag message changelog -->
<!-- See https://keepachangelog.com/en/1.0.0/ for examples -->

- Changed the `keychain::Balance` struct fields to use `Amount` instead of `u64`.
- Changed the `add_recipient` method on `TxBuilder` implementation to expect `bitcoin::Amount`.
- Changed the `sent_and_received`, and `net_value` methods on `KeyChainTxOutIndex` to expect `bitcoin::Amount`.
- Changed the `sent_and_received`, and `net_value` methods on `SpkTxOutIndex` to expect `bitcoin::Amount`.

### Checklists

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for the new feature
* [x] I've added docs for the new feature

#### Bugfixes:

* [x] This pull request breaks the existing API
* [ ] I've added tests to reproduce the issue which are now passing
* [x] I'm linking the issue being fixed by this PR

oleonardolima committed:
feat: update `wallet::Balance` to use `bitcoin::Amount`

- update all fields `immature`, ` trusted_pending`, `unstrusted_pending`
  and `confirmed` to use the `bitcoin::Amount` instead of `u64`
- update all `impl Balance` methods to use `bitcoin::Amount`
- update all tests that relies on `keychain::Balance`

oleonardolima commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -718,6 +718,7 @@ impl<'a, Cs: CoinSelectionAlgorithm> TxBuilder<'a, Cs, CreateTx> {
         self
     }
 
+    // TODO: (@leonardo) Should this expect/use `bitcoin::Amount` instead ? Would it be a huge breaking change ?
```
what do you folks think ? do you see any problem in tackling this change as well ?
cc @notmandatory @ValuedMammal @storopoli

oleonardolima replied:
_NOTE:_ the changes aren't hard (will do on separate commits anyway), it just impact some APIs

ValuedMammal replied:
yeah I'm open to having this change assuming we call `to_sat` before pushing onto recipients.

oleonardolima replied:
hmm, that's ok too, but why we can't we use Amount on the recipients, as well ?

storopoli replied:
Same, go for it.

ValuedMammal replied:
> why we can't we use Amount on the recipients, as well ?

I don't think it adds anything, because the library assumes all amounts are in satoshis and we should keep this invariant.

notmandatory replied:
@ValuedMammal do you mean not to use `Amount` for the `amount` param? I agree internally in `TxParams` it should stay sats but for the public API wouldn't `Amount` be safer?

notmandatory replied:
@oleonardolima yes thanks for taking this one on, it's one we'd still like to get into the 1.0 alpha.

ValuedMammal replied:
yes for changing this method `add_recipient` to accept an `Amount`. I don't think the type in `TxParams` needs to change in this PR

oleonardolima replied:
Good, I think we should update the `set_recipient` to accept `Amount` too, right ? @ValuedMammal @notmandatory wdyt ?

notmandatory replied:
Yes anywhere in the public API where we can replace a `u32` (or `i64`) representing sats with `Amount` (or `SignedAmount`) I'd like to do it it.

oleonardolima commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -270,6 +270,7 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
         self.spk_indices.get(script)
     }
 
+    // TODO: (@leonardo) Should this also be updated to return `(bitcoin::Amount, bitcoin::Amount)` instead of (u64, u64)
```
same as above, what do you folks think ? do you see any problem in tackling this change as well ?
cc @notmandatory @ValuedMammal @storopoli

storopoli replied:
I don't see an issue in changing this to `Amount`.

oleonardolima committed:
feat: use `Amount` on `spk_txout_index` and related

- update `wallet.rs` fns: `sent_and_received` fn
- update `keychain` `txout_index` fn: `sent_and_received and `net_value`

oleonardolima committed:
feat: use `Amount` on `TxBuilder::add_recipient`

oleonardolima commented:
I think it's ready for a new round of reviews

_I'll move the previous conversations to resolved after it, and if it's everything looks good on this new review round)_

storopoli commented:
> I think it's ready for a new round of reviews

Can you move it from Draft PR to a full PR?

oleonardolima commented:
> > I think it's ready for a new round of reviews
> 
> Can you move it from Draft PR to a full PR?

Yes, forgot to do it

oleonardolima commented:
~Oh, I forgot to run the tests on docs, working on it 😅~
Done!

storopoli commented on crates/bdk/src/wallet/coin_selection.rs:
```diff
@@ -92,7 +92,10 @@
+//!     builder.add_recipient(
+//!         to_address.script_pubkey(),
+//!         bitcoin::Amount::from_sat(50_000),
+//!     );
```
Isn't this covered in the

```rust
//! # use bitcoin::*;
```

on line 28?

```suggestion
//!     builder.add_recipient(
//!         to_address.script_pubkey(),
//!         Amount::from_sat(50_000),
//!     );
```

storopoli commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1175,7 +1175,7 @@ impl Wallet {
     ///    let mut builder =  wallet.build_tx();
     ///    builder
-    ///        .add_recipient(to_address.script_pubkey(), 50_000);
+    ///        .add_recipient(to_address.script_pubkey(), bitcoin::Amount::from_sat(50_000));
```
It's already imported in line 40

```suggestion
    ///        .add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000));
```

storopoli commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1557,7 +1557,7 @@ impl Wallet {
     ///     let mut builder = wallet.build_tx();
     ///     builder
-    ///         .add_recipient(to_address.script_pubkey(), 50_000)
+    ///         .add_recipient(to_address.script_pubkey(), bitcoin::Amount::from_sat(50_000))
```
Same

```suggestion
    ///         .add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000))
```

storopoli commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -1730,7 +1730,7 @@ impl Wallet {
     /// let mut psbt = {
     ///     let mut builder = wallet.build_tx();
-    ///     builder.add_recipient(to_address.script_pubkey(), 50_000);
+    ///     builder.add_recipient(to_address.script_pubkey(), bitcoin::Amount::from_sat(50_000));
```
Same

```suggestion
    ///     builder.add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000));
```

storopoli commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -29,7 +29,7 @@
 //! tx_builder
 //!     // Create a transaction with one output to `to_address` of 50_000 satoshi
-//!     .add_recipient(to_address.script_pubkey(), 50_000)
+//!     .add_recipient(to_address.script_pubkey(), bitcoin::Amount::from_sat(50_000))
```
Already imported in line 18

```suggestion
//!     .add_recipient(to_address.script_pubkey(), Amount::from_sat(50_000))
```

storopoli commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -94,8 +94,8 @@ impl TxBuilderContext for BumpFee {}
-///         .add_recipient(addr1.script_pubkey(), 50_000)
-///         .add_recipient(addr2.script_pubkey(), 50_000);
+///         .add_recipient(addr1.script_pubkey(), bitcoin::Amount::from_sat(50_000))
+///         .add_recipient(addr2.script_pubkey(), bitcoin::Amount::from_sat(50_000));
```
You are importing in L50

```suggestion
///         .add_recipient(addr1.script_pubkey(), Amount::from_sat(50_000))
///         .add_recipient(addr2.script_pubkey(), Amount::from_sat(50_000));
```

storopoli commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -104,7 +104,7 @@ impl TxBuilderContext for BumpFee {}
 ///     builder.ordering(TxOrdering::Untouched);
 ///     for addr in &[addr1, addr2] {
-///         builder.add_recipient(addr.script_pubkey(), 50_000);
+///         builder.add_recipient(addr.script_pubkey(), bitcoin::Amount::from_sat(50_000));
```
imported L50

```suggestion
///         builder.add_recipient(addr.script_pubkey(), Amount::from_sat(50_000));
```

storopoli commented on crates/bdk/src/wallet/tx_builder.rs:
```diff
@@ -274,7 +274,10 @@ impl<'a, Cs, Ctx> TxBuilder<'a, Cs, Ctx> {
-    ///     .add_recipient(to_address.script_pubkey(), 50_000)
+    ///     .add_recipient(
+    ///         to_address.script_pubkey(),
+    ///         bitcoin::Amount::from_sat(50_000),
```
imported L50

```suggestion
    ///         Amount::from_sat(50_000),
```

storopoli commented on crates/bdk/tests/psbt.rs:
```diff
@@ -1,5 +1,6 @@
-use bdk::bitcoin::{Amount, FeeRate, Psbt, TxIn};
+use bdk::bitcoin::{FeeRate, Psbt, TxIn};
 use bdk::{psbt, KeychainKind, SignOptions};
+use bitcoin::Amount;
```
Why this change?

oleonardolima replied:
I have no idea, probably rust-analyzer stuff 🤔, fixed it.

storopoli commented on crates/chain/src/keychain.rs:
```diff
@@ -90,26 +90,26 @@ impl<K> AsRef<BTreeMap<K, u32>> for ChangeSet<K> {
 
     /// Get the whole balance visible to the wallet.
-    pub fn total(&self) -> u64 {
+    pub fn total(&self) -> bitcoin::Amount {
```
Why not import `bitcoin::Amount` and reduce the clutter?

oleonardolima replied:
Done, much better now :)

storopoli commented on crates/chain/src/keychain/txout_index.rs:
```diff
@@ -285,7 +289,7 @@ impl<K: Clone + Ord + Debug> KeychainTxOutIndex<K> {
     ///
     /// [`sent_and_received`]: Self::sent_and_received
-    pub fn net_value(&self, tx: &Transaction, range: impl RangeBounds<K>) -> i64 {
+    pub fn net_value(&self, tx: &Transaction, range: impl RangeBounds<K>) -> SignedAmount {
```
Why this needs to be a `SignedAmount`?

oleonardolima replied:
It's because `net_value` used `i64` previously, and to keep the standard (and makes sense in my opinion given the  fn name) I used the `SignedAmount`, same for the other comments, they are all `net_value` references.

storopoli replied:
Oh I see...

storopoli commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -301,9 +305,11 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
     ///
     /// [`sent_and_received`]: Self::sent_and_received
-    pub fn net_value(&self, tx: &Transaction, range: impl RangeBounds<I>) -> i64 {
+    pub fn net_value(&self, tx: &Transaction, range: impl RangeBounds<I>) -> SignedAmount {
```
Why this needs to be a `SignedAmount`?

storopoli commented on crates/chain/src/tx_graph.rs:
```diff
@@ -1155,27 +1155,27 @@ impl<A: Anchor> TxGraph<A> {
+        let mut immature = bitcoin::Amount::ZERO;
+        let mut trusted_pending = bitcoin::Amount::ZERO;
+        let mut untrusted_pending = bitcoin::Amount::ZERO;
+        let mut confirmed = bitcoin::Amount::ZERO;
```
Why not import `bitcoin::Amount` and reduce the clutter?

storopoli commented on crates/chain/tests/test_spk_txout_index.rs:
```diff
@@ -20,14 +22,23 @@ fn spk_txout_sent_and_received() {
+        index.sent_and_received(&tx1, 1..),
+        (Amount::from_sat(0), Amount::from_sat(0))
+    );
+    assert_eq!(index.net_value(&tx1, ..), SignedAmount::from_sat(42_000));
```
Why this needs to be a `SignedAmount`?

storopoli commented on crates/chain/tests/test_spk_txout_index.rs:
```diff
@@ -53,10 +64,19 @@ fn spk_txout_sent_and_received() {
+        index.sent_and_received(&tx2, 1..),
+        (Amount::from_sat(0), Amount::from_sat(20_000))
+    );
+    assert_eq!(index.net_value(&tx2, ..), SignedAmount::from_sat(8_000));
```
Why this needs to be a `SignedAmount`?

storopoli reviewed (changes requested):
Some suggestions.

Most important, why we are using `SignedAmount` in some places?

oleonardolima commented:
> Some suggestions.
> 
> Most important, why we are using `SignedAmount` in some places?

It's because net_value used i64 previously, and to keep the standard (and makes sense in my opinion given the fn name) I used the SignedAmount, same for the other comments, they are all net_value references.

_(replying on the overall comment too)_

storopoli reviewed (approved):
ACK 362a67aefd5baea6a321721df05e9aff79894c86

Great work!

ValuedMammal commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -301,9 +305,11 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
-        received as i64 - sent as i64
+        (received - sent)
+            .to_signed()
+            .expect("Amount conversion to SignedAmount error")
```
I think you want to do `to_signed` on each Amount before doing the subtraction.

oleonardolima replied:
Yes, you're right, fixed it.

ValuedMammal commented on example-crates/example_cli/src/lib.rs:
```diff
@@ -506,7 +506,7 @@ where
             fn print_balances<'a>(
                 title_str: &'a str,
-                items: impl IntoIterator<Item = (&'a str, u64)>,
+                items: impl IntoIterator<Item = (&'a str, Amount)>,
```
This function intends to print out a value in sats, so for this to work we need another `to_sat` inside the function.

```diff
             fn print_balances<'a>(
                 title_str: &'a str,
                 items: impl IntoIterator<Item = (&'a str, Amount)>,
             ) {
                 println!("{}:", title_str);
                 for (name, amount) in items.into_iter() {
-                    println!("    {:<10} {:>12} sats", name, amount)
+                    println!("    {:<10} {:>12} sats", name, amount.to_sat())
```

oleonardolima replied:
Nice, good catch! Fixed it.

ValuedMammal commented on crates/chain/src/spk_txout_index.rs:
```diff
@@ -301,9 +305,14 @@ impl<I: Clone + Ord> SpkTxOutIndex<I> {
+        let signed_sent = sent.to_signed().expect(error_msg);
+        let signed_received = received.to_signed().expect(error_msg);
+
+        signed_received - signed_sent
```
```suggestion
        received.to_signed().expect("valid SignedAmount")
            - sent.to_signed().expect("valid SignedAmount")
```

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
````diff
@@ -978,7 +978,7 @@ impl Wallet {
     /// let (sent, received) = wallet.sent_and_received(tx);
     /// ```
-    pub fn sent_and_received(&self, tx: &Transaction) -> (u64, u64) {
+    pub fn sent_and_received(&self, tx: &Transaction) -> (Amount, Amount) {
````
In this case, the docs for this method shouldn't say that the returned amounts are "in satoshis".

oleonardolima replied:
Yes, updated it, thanks!

ValuedMammal commented on crates/bdk/tests/wallet.rs:
```diff
@@ -561,14 +561,15 @@ fn test_create_tx_change_policy_no_internal() {
     ));
 }
 
+// TODO: (@leonardo) Can we update this one to return an `bitcoin::Amount` instead ?
```
If you mean can we change `Wallet::calculate_fee` to return an `Amount`, I think possibly yes. But to avoid scope creep we should push it to a future PR.

```suggestion
```

oleonardolima replied:
Yes, that's what I mean. Cool, I'll address it in another PR.

ValuedMammal reviewed (commented):
Sorry my reviews have been intermittent. I think this is good.

The commit history could be cleaned up slightly:

- In df92624b011668216bba9908f6bd4fbb6db06457 we want to refer to `TxBuilder::add_recipient` (not CoinSelectionAlgorithm)

- Some changes in later commits (like fixing imports and wallet examples) could be squashed into earlier ones, but this not a blocker in my opinion.

Some other thoughts: I agree there is potentially more to do like updating `calculate_fee` and `Psbt::fee_amount` to use `Amount`. I'm still unsure how we should handle amounts internally. Honestly I don't mind leaving it as `u64`. If we decide that all amounts should be strongly typed, then `Amount` seems like a logical choice, my only concern is we should avoid the possibility of mixing up denominations.

oleonardolima commented:
> Sorry my reviews have been intermittent. I think this is good.
> 
> The commit history could be cleaned up slightly:
> 
> * In [df92624](https://github.com/bitcoindevkit/bdk/commit/df92624b011668216bba9908f6bd4fbb6db06457) we want to refer to `TxBuilder::add_recipient` (not CoinSelectionAlgorithm)
> * Some changes in later commits (like fixing imports and wallet examples) could be squashed into earlier ones, but this not a blocker in my opinion.
> 
> Some other thoughts: I agree there is potentially more to do like updating `calculate_fee` and `Psbt::fee_amount` to use `Amount`. I'm still unsure how we should handle amounts internally. Honestly I don't mind leaving it as `u64`. If we decide that all amounts should be strongly typed, then `Amount` seems like a logical choice, my only concern is we should avoid the possibility of mixing up denominations.

@ValuedMammal No worries, thanks for the comments and suggestions! I tried to address them, please let me know if something remains or if new needed changes appear.

ValuedMammal commented on crates/bdk/src/wallet/mod.rs:
```diff
@@ -950,10 +950,10 @@ impl Wallet {
     }
 
-    /// Compute the `tx`'s sent and received amounts (in satoshis).
+    /// Compute the `tx`'s sent and received amounts (in [`bitcoin::Amount`]).
```
What do you think of this?

```suggestion
    /// Compute the `tx`'s sent and received [`Amount`]s.
```

ValuedMammal reviewed (approved):
ACK 2e6aeb7499d9c4fadce4024fe26666976447ac2e

ValuedMammal commented:
Actually, changing `calculate_fee` and `Psbt::fee_amount` should be fairly easy to add to this PR but I'll let you decide if you want to tackle it here or later. Besides that, I'm not sure if there's still more places in the public API that would benefit from using `Amount`.

oleonardolima commented:
> Actually, changing `calculate_fee` and `Psbt::fee_amount` should be fairly easy to add to this PR but I'll let you decide if you want to tackle it here or later. Besides that, I'm not sure if there's still more places in the public API that would benefit from using `Amount`.

I updated `calculate_fee` and `fee_amount` on ac762f4a862db227432aa7daeeee424d2b653af3, but ended up touching other parts as well.

I would rather move this commit ac762f4a862db227432aa7daeeee424d2b653af3 to another PR, and keep only the previous ones here, wdyt ?

oleonardolima commented:
> > Actually, changing `calculate_fee` and `Psbt::fee_amount` should be fairly easy to add to this PR but I'll let you decide if you want to tackle it here or later. Besides that, I'm not sure if there's still more places in the public API that would benefit from using `Amount`.
> 
> I updated `calculate_fee` and `fee_amount` on [ac762f4](https://github.com/bitcoindevkit/bdk/commit/ac762f4a862db227432aa7daeeee424d2b653af3), but ended up touching other parts as well.
> 
> I would rather move this commit [ac762f4](https://github.com/bitcoindevkit/bdk/commit/ac762f4a862db227432aa7daeeee424d2b653af3) to another PR, and keep only the previous ones here, wdyt ?

I moved it to the other PR #1426, we can start discussing the changes on that one after this one is merged.

evanlinjin reviewed (approved):
ACK 22aa534d7648e5808414ea3adfcfb702572bd6c9
